src/chttpd/src/chttpd_util.erl

```diff
@@ -95,3 +99,68 @@ handle_legacy_config(OriginalHeaders, DefaultHeaderValue) ->
         false ->
             OriginalHeaders
     end.
+
+handle_request_helper(Req) ->
+    Args = couch_mrview_http:parse_params(Req, undefined),
+    ShardDbName = config:get("mem3", "shards_db", "_dbs"),
+    %% shard_db is not sharded but mem3:shards treats it as an edge case
+    %% so it can be pushed thru fabric
+    {ok, Info} = fabric:get_db_info(ShardDbName),
+    Etag = couch_httpd:make_etag({Info}),
+    Options = [{user_ctx, Req#httpd.user_ctx}],
+    {ok, Resp} = chttpd:etag_respond(Req, Etag, fun() ->
+        {ok,Resp}=chttpd:start_delayed_json_response(Req,200,[{"ETag",Etag}]),
+        VAcc = #vacc{req=Req,resp=Resp},
+        fabric:all_docs(ShardDbName, Options, fun callback_func/2, VAcc, Args)
+    end),
+    case is_record(Resp, vacc) of
+        true -> {ok, Resp#vacc.resp};
+        _ -> {ok, Resp}
+    end.
+
+callback_func({meta, _Meta}, #vacc{resp=Resp0}=Acc) ->
+    {ok, Resp1} = chttpd:send_delayed_chunk(Resp0, "["),
+    {ok, Acc#vacc{resp=Resp1}};
+callback_func({row, Row}, #vacc{resp=Resp0}=Acc) ->
+    [PathParts] = Acc#vacc.req#httpd.path_parts,
+    Prepend = couch_mrview_http:prepend_val(Acc),
+    case ?b2l(PathParts) of
+        "_all_dbs" ->
+            case couch_util:get_value(id, Row) of <<"_design", _/binary>> ->
+                {ok, Acc};
+                DbName ->
+                    {ok, Resp1} = chttpd:send_delayed_chunk(
+                        Resp0, [Prepend, ?JSON_ENCODE(DbName)]),
+                    {ok, Acc#vacc{prepend=",", resp=Resp1}}
+            end;
+        "_dbs_info" ->
+            DbName = couch_util:get_value(id, Row),
+            DbInfo = get_dbs_info_shards(DbName),
+            Chunk = [Prepend, ?JSON_ENCODE({[{key,DbName}, {info,{DbInfo}}]})],
+            {ok, Resp1} = chttpd:send_delayed_chunk(Resp0, Chunk),
+            {ok, Acc#vacc{prepend = ",", resp = Resp1}}
+    end;
+callback_func(complete, #vacc{resp=Resp0}=Acc) ->
+    {ok, Resp1} = chttpd:send_delayed_chunk(Resp0, "]"),
+    {ok, Resp2} = chttpd:end_delayed_json_response(Resp1),
+    {ok, Acc#vacc{resp=Resp2}};
+callback_func({error, Reason}, #vacc{resp=Resp0}=Acc) ->
+    {ok, Resp1} = chttpd:send_delayed_error(Resp0, Reason),
+    {ok, Acc#vacc{resp=Resp1}}.
+
+get_dbs_info_shards(DbName) ->
+    Timeout = fabric_util:request_timeout(),
+    IsolatedFun = fun() -> fabric:get_db_info(DbName) end,
+    try fabric_util:isolate(IsolatedFun, Timeout) of
+        {ok, DbsInfo} ->
+            DbsInfo;
+        {error, Error} ->
+            ErrMsg = "~p : could not get _dbs_info for Db:~p Error:~p",
+            couch_log:error(ErrMsg, [?MODULE, DbName, Error]),
+            []
+    catch
+        _Tag:Error ->
+            ErrMsg = "~p : could not get _dbs_info for Db:~p Error:~p",
+            couch_log:error(ErrMsg, [?MODULE, DbName, Error]),
+            []
+    end.
```

Comment:
It could be that this may be the only function we could justify keeping in this module as there is at least another place where we `fabric:get_db_info/1` in the context of another request. The handler and callbacks just don't seem generic enough to be helpers.

The only thing we want to think about is what should happen during errors. If we return `[]` so it would seem `_dbs_info` response would return a `{"key": DbName, "info": {}}` row which I am not sure is quite right.

Reply:
Convert [] to `{error, not_found}`.
Thanks for your advice!